### DISCUSSION TITLE
Mechanism for call hierarchy  Rexx linkage, native linkage and different platforms

### DISCUSSION THREAD

**Discussion by rvjansen** [adesutherland/CREXX#75] [Ideas]
Two leading principles, performance and portability

Can we use the namespace/module mechanism for choosing which version of external functions (like bif's) for Rexx linkage we employ?

Platform specific native functionality needs to be compiled in, are we using #defines , directories etc?

Can we classify C functions/rxas instructions in Core, Performance, etc? We can then have an as small as possible kernel, use the module system (in options perhaps) to enable those, for specific performance related implementations, and have a fallback mechanism to fall back on a generic Rexx implementation.

**Comment by rvjansen** [comment]
Some more thoughts. This is based on the observation that I could have oldsubstr.rxas exposing an entrypoint of substr and being perfectly exchangeable in executions of rxvm testsubstr.
So if we have, for every library, a hierarchy of:
- zero
- rxas
- native
then we can call the function at each level - the rexx and rxas versions using crexx rxas linkage, and the native versions also, *if we implement an rxas instruction that is platform dependent*  (instruction set architecture, os) and package the native functions in an rxas module that uses that instruction and marshalls arguments and return addresses. Every os/platform version has only one version of that instruction, which can be named, e.g. callnative.

**Comment by adesutherland** [comment]
We should certainly have a nativecall instruction, and I suspect the format of the arguments and return will end up a bit like rexxsaa stuff, but not binary compatible. Realistically if we can get an existing library to compile will minimal changes we should count that as success.

However I don't see the need of any of the standard bifs being native given performance findings?

**Reply by rvjansen** [reply]
True, not yet, but I think we should keep an open mind, for example for the emulated platforms

**Reply by rvjansen** [reply]
My thought was to have a native call instruction for calling native code with rexx register conventions, and build the external library/adress calls on top of that.

**Reply by rvjansen** [reply]
So we call it as if it is crexx, with a0 a1 etc, and per platform/os/ linking convention we have native implementations, so we can do register or stack linkage, fastcall, pascal etc. Convention to be read from a register as specified by the caller
